src/main/kotlin/no/nav/syfo/narmestelederrelasjon/kafka/KafkaNarmesteLederRelasjon.kt

```diff
@@ -0,0 +1,101 @@
+package no.nav.syfo.narmestelederrelasjon.kafka
+
+import com.fasterxml.jackson.module.kotlin.readValue
+import kotlinx.coroutines.*
+import no.nav.syfo.application.ApplicationState
+import no.nav.syfo.application.Environment
+import no.nav.syfo.application.database.DatabaseInterface
+import no.nav.syfo.narmestelederrelasjon.database.NoElementInsertedException
+import no.nav.syfo.narmestelederrelasjon.database.createNarmesteLederRelasjon
+import no.nav.syfo.narmestelederrelasjon.kafka.domain.NarmesteLederLeesah
+import no.nav.syfo.util.*
+import org.apache.kafka.clients.consumer.ConsumerRecords
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.time.Duration
+
+private val log: Logger = LoggerFactory.getLogger("no.nav.syfo.narmestelederrelasjon.kafka")
+
+const val NARMESTE_LEDER_RELASJON_TOPIC = "teamsykmelding.syfo-narmesteleder-leesah"
+
+fun CoroutineScope.createListenerNarmesteLederRelasjon(
+    applicationState: ApplicationState,
+    action: suspend CoroutineScope.() -> Unit
+): Job =
+    launch {
+        try {
+            action()
+        } finally {
+            applicationState.alive = false
+        }
+    }
+
+fun blockingApplicationLogicNarmesteLederRelasjon(
+    applicationState: ApplicationState,
+    database: DatabaseInterface,
+    environment: Environment,
+) {
+    log.info("Setting up kafka consumer NarmesteLederLeesah")
+
+    val consumerProperties = kafkaNarmesteLederRelasjonConsumerConfig(environment.kafka)
+    val kafkaConsumerNarmesteLederRelasjon = KafkaConsumer<String, String>(consumerProperties)
+
+    kafkaConsumerNarmesteLederRelasjon.subscribe(
+        listOf(NARMESTE_LEDER_RELASJON_TOPIC)
+    )
+    if (environment.toggleKafkaProcessingEnabled) {
+        while (applicationState.ready) {
```

Comment:
Det er mulig ready fortsatt er false når koden kommer hit, og da vil jo ingenting starte opp.

Så kanskje bedre å bruke applicationState.alive i stedet? Det eneste potensielle problemet med det er at databasen kanskje ikke er ferdig initialisert.

Reply:
Mulig jeg har misforstått noe, men om oppretting av database i denne blokken gjøres sekvensielt så vil database være koblet opp før man kommer til KafkaModule:
```
 module {
                databaseModule(
                    environment = environment,
                )
                apiModule(
                    applicationState = applicationState,
                    database = applicationDatabase,
                    environment = environment,
                    wellKnownInternalAzureAD = wellKnownInternalAzureAD,
                )
                kafkaModule(
                    applicationState = applicationState,
                    database = applicationDatabase,
                    environment = environment,
                )
            }
```

Reply:
Ja, sannsynligvis går det bra med while (applicationState.alive)

Reply:
@geir-waagboe 
PS: nå lagde jeg en ny commit hvor lateinit av database erstattes med init av database før modules.

Reply:
OK, fint. Men jeg mener fortsatt at du bør bytte ut 
while (applicationState.ready) 
med 
while (applicationState.alive)
for ready settes ikke til true før hele initialiseringen er ferdig (og skal heller ikke settes før - ellers kan pod'en få rutet trafikk til seg før den er klar til å prosessere)

Reply:
Ja det er sant, men da blir jo hele rekkefølgen "feil" i begge varianter.
Jeg skal se etter en løsning som løser begge deler.